### PR TITLE
Make AsyncSearchResponse ref-counted

### PR DESCRIPTION
**Pull request by original-brownbear**
It's in the title, this class references `SearchResponse`, so it itself must become ref-counted.

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/docs

**Comment by original-brownbear** [comment]
Buildkite test this

**Comment by original-brownbear** [comment]
@elasticmachine update branch

**Comment by quux00** [comment]
I see that `AsyncStatusResponse` has not been made `RefCounted`, because it does not hold a reference to a `SearchResponse`. However, it does get a reference to `SearchResponse.Clusters` (from MutableSearchResponse). That Clusters object could be part of a SearchResponse that gets ref-counted to zero. Is that going to be a problem? Should we split the Clusters object out to its own class?

File: x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/CrossClusterAsyncSearchIT.java [hunk 1]
```diff
@@ -413,40 +423,44 @@ public void testClusterDetailsAfterCCSWithFailuresOnAllShards() throws Exception
...
-
-        {
-            AsyncSearchResponse finishedResponse = getAsyncSearch(response.getId());
-            assertTrue(finishedResponse.isPartial());
+        final AsyncSearchResponse response = submitAsyncSearch(request);
+        try {
+            assertNotNull(response.getSearchResponse());
+            waitForSearchTasksToFinish();
+            final AsyncSearchResponse finishedResponse = getAsyncSearch(response.getId());
+            try {
```
Comment (quux00) [inline x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/CrossClusterAsyncSearchIT.java:431]: This double nesting of try/finally blocks seems unnecessary - I don't think you've had to do it in other tests. What if you cache the response.getId() like elsewhere and then `response.decRef` before you start the second try/finally?

Comment (original-brownbear) [reply x-pack/plugin/async-search/src/internalClusterTest/java/org/elasticsearch/xpack/search/CrossClusterAsyncSearchIT.java]: ++ indeed, that was just me getting tired I think. Much better this way, thanks for spotting!

**Comment by original-brownbear** [comment]
Thanks Michael all points addressed I think/hope! :)

**Review by quux00 — APPROVED** [review]
LGTM. Nice work on this!

**Pull request merged**